File: windows/JVoice.Core/Models/SettingsState.cs
namespace JVoice.Core.Models;

public sealed record SettingsState(
    int SchemaVersion,
    ToneStyle Mode,
    WhisperModelOption Model,
    TranscriptionLanguage Language,
    IReadOnlyList<string> CustomWords,
    bool RemoveFillerWords,
    IReadOnlyList<CorrectionRule> Corrections,
    HotkeyChord Hotkey,
    GameDetectionMode GameMode = GameDetectionMode.Balanced,
    bool DeveloperTerms = true,
    // ── v3 (Windows-only dictation features; no macOS counterpart) ──
    // Paste target: when true, dictation copies to the clipboard instead of auto-pasting.
    bool CopyToClipboardOnly = false,
    // Opt-in "undo last paste" chord. null = disabled (any registered global chord is swallowed
    // system-wide, so there is deliberately no default — the user assigns a rare chord).
    HotkeyChord? UndoHotkey = null,
    // Whisper "translate" task: output English regardless of the spoken/source Language.
    bool TranslateToEnglish = false,
    // Master toggle for app-aware modes (auto-switch tone by foreground app; built-in code apps).
    bool AppAwareModes = true,
    // User per-app rules (built-in code apps are implicit in AppModeResolver, not persisted).
    // Nullable param + normalized property so a defaulted/positional construction never yields null.
    IReadOnlyList<AppModeRule>? AppModeRules = null,
    // ── v4 (Windows-only) ──
    // Automatically check GitHub for a newer JVoice release and offer an in-app update. On by
    // default; the check is a single anonymous GET to the GitHub API that sends no user data (the
    // second and only other network call besides the one-time model download). No macOS counterpart.
    bool CheckForUpdates = true,
    // ── v5 (Windows-only) ──
    // The capture endpoint to record from. null = follow the system default (the behavior before
    // v5, and still the default). Stored because the system default is NOT always a real
    // microphone: virtual devices (Voicemod, NVIDIA Broadcast, VB-Cable, Elgato Wave Link, OBS)
    // routinely own the default and emit digital silence when their app isn't running, and
    // changing the OS-wide default to work around that would break the apps that want it (§7 #46).
    string? InputDeviceId = null,
    // Friendly name of InputDeviceId, cached purely so Settings and error messages can name the
    // device without enumerating, and so a re-plugged device is still recognizable to the user.
    // Never used for matching — InputDeviceId is the identity.
    string? InputDeviceName = null,
    // ── v6 (Windows-only) ──
    // Opt-out toggle for spoken-mathematics notation ("x squared plus 1" → "x² + 1"). Windows-only;
    // no macOS counterpart. Default ON like DeveloperTerms: MathSpeech only rewrites a run of words
    // when an activating construct found its operands, so ordinary speech comes back byte-identical
    // and the feature is invisible until someone actually dictates mathematics.
    bool MathNotation = true)
{
    public const int CurrentSchemaVersion = 6;

    public IReadOnlyList<AppModeRule> AppModeRules { get; init; } = AppModeRules ?? Array.Empty<AppModeRule>();

    public static SettingsState Default => new(
        SchemaVersion: CurrentSchemaVersion,
        Mode: ToneStyle.Casual,
        // Windows-only divergence from macOS (which defaults to .tiny — see
        // ../../../Sources/JVoice/Models/SettingsState.swift). On Windows the default is the
        // most-accurate model: with GPU acceleration Large ("large-v3-turbo") is both the best
        // and fast, so it is the right out-of-the-box choice. The Settings "Whisper Model" card
        // shows a warning advising the user to keep this on Large unless they know they need a
        // smaller model. Kept in sync with SettingsStateJson.ParseModel's fallback so any
        // unspecified-model path lands on the same default.
        Model: WhisperModelOption.LargeTurbo,
        Language: TranscriptionLanguage.English,
        CustomWords: Array.Empty<string>(),
        RemoveFillerWords: true,
        Corrections: Array.Empty<CorrectionRule>(),
        // Windows-only: the global hotkey chord. On macOS the KeyboardShortcuts library
        // persists this in its own UserDefaults key, outside SettingsState; the Windows
        // port folds it into settings.json like the other Windows-only fields. Default is
        // Ctrl+Shift+Space (HotkeyChord.Default).
        Hotkey: HotkeyChord.Default,
        GameMode: GameDetectionMode.Balanced,
        DeveloperTerms: true,
        CopyToClipboardOnly: false,
        UndoHotkey: null,
        TranslateToEnglish: false,
        AppAwareModes: true,
        AppModeRules: Array.Empty<AppModeRule>(),
        CheckForUpdates: true,
        // null = follow the system default capture endpoint (v5).
        InputDeviceId: null,
        InputDeviceName: null,
        // Spoken-mathematics notation ON by default (v6) — opt-out, like DeveloperTerms.
        MathNotation: true);
}

File: Sources/JVoice/Models/SettingsState.swift
import Foundation

public struct SettingsState: Codable, Equatable, Sendable {
    public static let currentSchemaVersion: Int = 4
    public var schemaVersion: Int = SettingsState.currentSchemaVersion
    public var mode: AppMode
    public var model: WhisperModelOption
    public var language: TranscriptionLanguage
    public var customWords: [String]
    public var removeFillerWords: Bool
    public var theme: AppTheme
    /// v3 dictation-parity fields (mirror the Windows port's SettingsState v3).
    public var developerTerms: Bool
    public var translateToEnglish: Bool
    public var copyToClipboardOnly: Bool
    public var appAwareModes: Bool
    public var appModeRules: [AppModeRule]
    /// v4: spoken mathematics → real notation ("x squared equals 4" → "x² = 4").
    /// Opt-out, like the Windows port's `MathNotation` (schema v6 there).
    public var mathNotation: Bool

    public var whisperModel: WhisperModelOption {
        get { model }
        set { model = newValue }
    }

    public init(
        mode: AppMode = .casual,
        model: WhisperModelOption = .tiny,
        language: TranscriptionLanguage = .english,
        customWords: [String] = [],
        removeFillerWords: Bool = true,
        theme: AppTheme = .dark,
        developerTerms: Bool = true,
        translateToEnglish: Bool = false,
        copyToClipboardOnly: Bool = false,
        appAwareModes: Bool = false,
        appModeRules: [AppModeRule] = [],
        mathNotation: Bool = true
    ) {
        self.mode = mode
        self.model = model
        self.language = language
        self.customWords = customWords
        self.removeFillerWords = removeFillerWords
        self.theme = theme
        self.developerTerms = developerTerms
        self.translateToEnglish = translateToEnglish
        self.copyToClipboardOnly = copyToClipboardOnly
        self.appAwareModes = appAwareModes
        self.appModeRules = appModeRules
        self.mathNotation = mathNotation
    }

    private enum CodingKeys: String, CodingKey {
        case schemaVersion
        case mode
        case model
        case language
        case customWords
        case removeFillerWords
        case theme
        case developerTerms
        case translateToEnglish
        case copyToClipboardOnly
        case appAwareModes
        case appModeRules
        case mathNotation
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let version = try container.decodeIfPresent(Int.self, forKey: .schemaVersion) ?? 0
        self.schemaVersion = SettingsState.currentSchemaVersion   // always normalize forward
        if version > SettingsState.currentSchemaVersion {
            throw DecodingError.dataCorruptedError(
                forKey: .schemaVersion,
                in: container,
                debugDescription: "Settings written by a newer JVoice build (v\(version) > v\(SettingsState.currentSchemaVersion)). Refusing to read."
            )
        }
        mode = (try? container.decode(AppMode.self, forKey: .mode)) ?? .casual
        model = (try? container.decode(WhisperModelOption.self, forKey: .model)) ?? .tiny
        language = try container.decodeIfPresent(TranscriptionLanguage.self, forKey: .language) ?? .english
        customWords = try container.decodeIfPresent([String].self, forKey: .customWords) ?? []
        removeFillerWords = try container.decodeIfPresent(Bool.self, forKey: .removeFillerWords) ?? true
        theme = try container.decodeIfPresent(AppTheme.self, forKey: .theme) ?? .dark
        // v3 fields: absent in v1/v2 blobs → default (developerTerms ON, rest OFF).
        developerTerms = try container.decodeIfPresent(Bool.self, forKey: .developerTerms) ?? true
        translateToEnglish = try container.decodeIfPresent(Bool.self, forKey: .translateToEnglish) ?? false
        copyToClipboardOnly = try container.decodeIfPresent(Bool.self, forKey: .copyToClipboardOnly) ?? false
        appAwareModes = try container.decodeIfPresent(Bool.self, forKey: .appAwareModes) ?? false
        appModeRules = try container.decodeIfPresent([AppModeRule].self, forKey: .appModeRules) ?? []
        // v4 field: absent in v1–v3 blobs → default ON, so an upgrade gains the feature.
        mathNotation = try container.decodeIfPresent(Bool.self, forKey: .mathNotation) ?? true
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(SettingsState.currentSchemaVersion, forKey: .schemaVersion)
        try container.encode(mode, forKey: .mode)
        try container.encode(model, forKey: .model)
        try container.encode(language, forKey: .language)
        try container.encode(customWords, forKey: .customWords)
        try container.encode(removeFillerWords, forKey: .removeFillerWords)
        try container.encode(theme, forKey: .theme)
        try container.encode(developerTerms, forKey: .developerTerms)
        try container.encode(translateToEnglish, forKey: .translateToEnglish)
        try container.encode(copyToClipboardOnly, forKey: .copyToClipboardOnly)
        try container.encode(appAwareModes, forKey: .appAwareModes)
        try container.encode(appModeRules, forKey: .appModeRules)
        try container.encode(mathNotation, forKey: .mathNotation)
    }
}
